src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,1008 @@
+      }
+    }
+  }
+
+  /**
+   * Treats minimal standard deviations and numeric problems. Increases sigma.
+   */
+  template<typename funcType>
+  void CMAES<funcType>::testMinStdDevs(void)
+  {
+    if(!this->rgDiffMinChange)
+      return;
+
+    for(int i = 0; i < N; ++i)
+      while(this->sigma*std::sqrt(this->C[i][i]) < this->rgDiffMinChange[i])
+        this->sigma *= std::exp(double(0.05) + this->cs / this->damps);
+  }
+
+  /**
+   * Adds the mutation sigma*B*(D*z).
+   * @param x Search space vector.
+   * @param eps Mutation factor.
+   */
+  template<typename funcType>
+  void CMAES<funcType>::addMutation(double* x, double eps)
+  {
+    for(int i = 0; i < N; ++i)
+      tempRandom[i] = rgD[i]*rand.gauss();
+    for(int i = 0; i < N; ++i)
+    {
+      double sum = 0.0;
+      for(int j = 0; j < N; ++j)
+        sum += B[i][j]*tempRandom[j];
+      x[i] = xmean[i] + eps*sigma*sum;
+    }
+  }
+
+  /**
+   * Initializes the CMA-ES algorithm.
+   * @param parameters The CMA-ES parameters in the parameters.h file
+   * @return Array of size lambda that can be used to assign fitness values and
+   *         pass them to updateDistribution()
+   */
+  template<typename funcType>
+  double* CMAES<funcType>::init()
+  {
+
+    stopMessage = "";
+
+    double trace(0);
+    for(int i = 0; i < N; ++i)
+      trace += rgInitialStds[i]*rgInitialStds[i];
+    sigma = std::sqrt(trace/N);
+
+    chiN = std::sqrt((double) N) * (double(1) - double(1)/(double(4)*N) + double(1)/(double(21)*N*N));
+    eigensysIsUptodate = true;
+    doCheckEigen = false;
+    genOfEigensysUpdate = 0;
+
+    double dtest;
+    for(dtest = double(1); dtest && dtest < double(1.1)*dtest; dtest *= double(2))
+      if(dtest == dtest + double(1))
+        break;
+    dMaxSignifKond = dtest / double(1000);
+
+    gen = 0;
+    countevals = 0;
+    state = INITIALIZED;
+    dLastMinEWgroesserNull = double(1);
+
+    pc = new double[N];
+    ps = new double[N];
+    tempRandom = new double[N+1];
+    BDz = new double[N];
+    xmean = new double[N+2];
+    xmean[0] = N;
+    ++xmean;
+    xold = new double[N+2];
+    xold[0] = N;
+    ++xold;
+    xBestEver = new double[N+3];
+    xBestEver[0] = N;
+    ++xBestEver;
+    xBestEver[N] = std::numeric_limits<double>::max();
+    output = new double[N+2];
+    output[0] = N;
+    ++output;
+    rgD = new double[N];
+    C = new double*[N];
+    B = new double*[N];
+    publicFitness = new double[lambda];
+    functionValues = new double[lambda+1];
+    functionValues[0] = lambda;
+    ++functionValues;
+    const int historySize = 10 + (int) ceil(3.*10.*N/lambda);
+    funcValueHistory = new double[historySize + 1];
+    funcValueHistory[0] = (double) historySize;
+    funcValueHistory++;
+
+    for(int i = 0; i < N; ++i)
+    {
+      C[i] = new double[i+1];
+      B[i] = new double[N];
+    }
+    index = new int[lambda];
+    for(int i = 0; i < lambda; ++i)
+        index[i] = i;
+    population = new double*[lambda];
```

Comment:
Use ``arma::mat`` here. Btw. we can avoid the usage of the first index to store the size, because we can just use ``population.n_rows`` and ``population.n_cols``.